Precalculate mod scores in periodic celery background task

DasSkelett opened the pull request:
## Problem
All routes that use `search_mods()` in one way or another are incredibly slow on production.
The cause of this is that on every request, we fetch (nearly*) _all_ mods from the database, calculate the mods's scores, sort the list by the score, and afterwards paginate it.
*: we only filter by `mod.published` and, if given, the search terms before.

This affects the following routes (via `get_mods()`):
* `/browse/all`
* `/browse/top`
* `/<gameshort>/browse/all`
* `/<gameshort>/browse/top`
* `/search`
* `/<gameshort>/search`
* `/api/browse/top`
Directly calling `search_mods()`:
* `/browse`
* `/<gameshort>/browse`
* `/api/search/mod`

## Solution
I implemented a new background task using celery, which seemed to be the most straightforward solution without getting into trouble with parallel gunicorn instances.
It's based on this documentation: https://docs.celeryproject.org/en/stable/userguide/periodic-tasks.html
For the celery scheduler to start, I had to append `-B` parameter to the systemd + docker scripts startup scripts. This is not recommended by production use, so maybe we have create a separate systemd unit for it.
> You can also embed beat inside the worker by enabling the workers -B option, this is convenient if you’ll never run more than one worker node, but it’s not commonly used and for that reason isn’t recommended for production use

`celery.calculate_mod_scores()` runs every 24 hours (set up in `celery.setup_periodic_tasks()`), fetches all mods, calculates the score for them and writes them to `Mod.score`. This will take around 10 seconds on prod I think (based on the current request time of the above listed routes).
`Mod.score` is a new database column of type `float`.

We also recalculate the score whenever we change a property of a mod that affects it.

`search.weigh_result()`, renamed to `search.get_mod_score()`, doesn't consider the number of matching search terms anymore, since they are not known at computation time.
We may or may not want to replace this functionality in one way or another, to show better matching mods further up.

`search.search_mods()` is updated to simply order the `Mod` table by `score`, and to use proper database-side pagination with `.limit()` and `.offset()`.
It also combines "special" search term filters (with special string prefixes) with `and_` instead of `or_` now, which IMHO makes more sense.
I considered doing something similar for the regular search terms, which would solve the aforementioned problem, but I have to think of a way to solve this elegantly.

HebaruSan commented on KerbalStuff/search.py:
```diff
@@ -19,68 +18,64 @@ def weigh_result(result, terms):
+    query = query.order_by(desc(Mod.score))
+
+    total_pages = math.ceil(query.count() / limit)
+    print(page)
```
Left behind debug statement?

DasSkelett replied:
Yes, already removed in local branch shortly after opening the PR, just didn't push it yet ^^

HebaruSan commented on KerbalStuff/celery.py:
```diff
@@ -84,6 +87,18 @@ def update_from_github(working_directory, branch, restart_command):
 
+@app.on_after_configure.connect
+def setup_periodic_tasks(sender, **kwargs):
+    sender.add_periodic_task(600, calculate_mod_scores.s(), name='calculate mod scores')
```
There are two score components that depend on time.

- 100 point bonus for creation within last 30 days
- 1 point penalty per 5 days since last update, up to a limit of 100 points

Given this, a 10 minute interval is extreme overkill. We could do 5 days and expect it to work fine.

HebaruSan commented:
Could we refresh a mod's individual score at create, edit, update, follow/unfollow, and download? This would help with any hiccups that might come with a longer celery timer interval.

DasSkelett commented:
> There are two score components that depend on time.
> 
>     * 100 point bonus for creation within last 30 days
> 
>     * 1 point penalty per 5 days since last update, up to a limit of 20 points
> 
> 
> Given this, a 10 minute interval is extreme overkill. We could do 5 days and expect it to work fine.

> Could we refresh a mod's individual score at create, edit, update, follow/unfollow, and download? This would help with any hiccups that might come with a longer celery timer interval.

The first suggestion depends on the second one. Because while all the other components don't depend on the time, they still change constantly. And it would take up to 5 days for any changes to reflect in the score, for example new mods would stay at 0 points this whole time.
We could change the default value to 100 for example, but this would still be missing all the other potential bonuses.

So if we want to increase the interval between the the calculations, we would have to also do your second suggestion.
However, for one I don't think there's a real need. I estimate the calculation to take around 6 seconds based on the page load times, maybe less since we removed the search-term part.
This wouldn't be a huge deal to do every 10 minutes (we don't depend on any AWS burst credits or GitHub API limits :wink: ).

Furthermore, we would have to call the score refresh everywhere we save changes that would affect the mod score. Not a huge deal, but we'd have to remember adding or removing calls whenever we update the score algorithm.

HebaruSan commented:
> while all the other components don't depend on the time, they still change constantly. And it would take up to 5 days for any changes to reflect in the score

True if you think of the exact score values as important, but their true functional role is as relative **rankings**. What matters is not the exact score that a mod has, but where it falls in the big sort of all mods; a set of old un-updated scores is still "accurate enough" if the up-to-date values would still sort those mods in the same order. Typically the mods within a given range of scores will "move together" as they all get a similar number of downloads here and there over time, so it's going to be rare for the rankings to change within any 5-day period. The main exception would be when some new mod "takes off" in popularity and starts surging ahead, and updating the score in the download route would cover that.

> I don't think there's a real need

I guess I see the inverse—we don't _need_ to refresh every score every 10 minutes, because the new score-rankings will be the same as the old ones almost every time (very hard to estimate without access to the prod db, but... maybe a mod changes rank every few days?). Since this scheduled task is new and we don't know what problems might come up for SpaceDock in the future, it's good to avoid taking up unnecessary system resources when we can.

HebaruSan commented on KerbalStuff/objects.py:
```diff
@@ -165,6 +165,7 @@ class Mod(Base):
     external_link = Column(String(512))
     license = Column(String(128))
     votes = Column(Integer, default=0)
+    score = Column(Float, default=0, nullable=False)
```
I think we can make the db side of our query faster by adding `index=True` here. This would allow the db to essentially pre-store the rows in sorted order so when we say `order_by().offset().limit()`, it could jump directly to those rows without having to access the entire table and sort.

- https://www.postgresql.org/docs/current/indexes-ordering.html

RockyTV commented:
Why not have the score update every day? I don't think it would take much time and most times the changes won't be noticeable by users if we update every 10 minutes. We could set the script to run every day at a time interval where we have the least access.

HebaruSan commented:
Another consideration... We currently have two pieces of functionality, emails and auto-restarts, reliant on Celery that are inexplicably not working. It would be good to have a functional fallback if we are going to put another egg in that basket.

DasSkelett commented:
I've set it to 86400 seconds == 1 day for now, which sounds like a good compromise.
I also like the idea of running it at a set time of day, more cron-job like, like @RockyTV suggested, I'll look into that.

I also added score updates to all places in the code where we change a property of a mod that affects the score.
I'm not 100% sure I placed it correctly in `api.create_mod()`, I'll have to test it.

Furthermore we also calculate all the scores during the database migration now, so we start with a correctly filled table.

DasSkelett commented:
The last commit fixes the score calculation in `create_mod()` by moving it below the `db.commit()` and calling a second `db.commit()`.

I also think I finally found a good solution to the search problem, see the code comments.
I'm not sure if it's a good idea to search the whole mod description or if that is too performance-heavy. Feedback welcome.

DasSkelett commented:
Okay, rebased on alpha for #281, adjusted the `down_revision` and the date, so the files sort in the same order as the revision chain.
From my side it's ready to merge!
